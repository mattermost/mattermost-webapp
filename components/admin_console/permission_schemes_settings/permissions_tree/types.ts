// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

export type Permission = {
    id: string;
    combined?: boolean;
    permissions: string[];
}
export type Group = {
    id: string;
    permissions: Array<Permission | string>;
}

export type AdditionalValues = {
    edit_post: {
        editTimeLimitButton: JSX.Element;
    };
}
