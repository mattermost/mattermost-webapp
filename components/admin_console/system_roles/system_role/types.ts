// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

export type ReadAccess = 'read';
export const readAccess: ReadAccess = 'read';
export type WriteAccess = 'write';
export const writeAccess: WriteAccess = 'write';
export type NoAccess = false;
export const noAccess: NoAccess = false;
export type MixedAccess = 'mixed';
export const mixedAccess: MixedAccess = 'mixed';

export type PermissionAccess = ReadAccess | WriteAccess | NoAccess;
export type PermissionsToUpdate = Record<string, ReadAccess | WriteAccess | NoAccess>;
export type PermissionToUpdate = {
    name: string;
    value: PermissionAccess;
};

export type SystemSection = {
    name: string;
    hasDescription?: boolean;
    subsections?: SystemSection[];
    disabled?: boolean;
}
