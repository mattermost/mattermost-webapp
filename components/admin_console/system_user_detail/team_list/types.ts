// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {Team, TeamMembership} from '@mattermost/types/teams';

export type TeamWithMembership = Team & TeamMembership;
