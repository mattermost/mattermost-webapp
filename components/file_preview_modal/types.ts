// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {FileInfo} from '@mattermost/types/files';

export type LinkInfo = {
    has_preview_image: boolean;
    link: string;
    extension: string;
    name: string;
}

export function isFileInfo(info: FileInfo | LinkInfo): info is FileInfo {
    return Boolean((info as FileInfo).id);
}
