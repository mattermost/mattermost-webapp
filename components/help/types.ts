// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

export enum HelpLink {
    Messaging = 'Messaging',
    Composing = 'Composing',
    Mentioning = 'Mentioning',
    Formatting = 'Formatting',
    Attaching = 'Attaching',
    Commands = 'Commands',
}
