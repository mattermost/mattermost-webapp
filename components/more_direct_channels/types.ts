// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {Channel} from '@mattermost/types/channels';
import {UserProfile} from '@mattermost/types/users';

import {Value} from 'components/multiselect/multiselect';

// Not to be confused with the GroupChannel type used for LDAP groups
export type GroupChannel = Channel & {
    profiles: UserProfile[];
}

export function isGroupChannel(option: UserProfile | GroupChannel): option is GroupChannel {
    return (option as GroupChannel)?.type === 'G';
}

export type Option = (UserProfile & {last_post_at?: number}) | GroupChannel;

export type OptionValue = Option & Value;

export function optionValue(option: Option): OptionValue {
    return {
        value: option.id,
        label: isGroupChannel(option) ? option.display_name : option.username,
        ...option,
    };
}
