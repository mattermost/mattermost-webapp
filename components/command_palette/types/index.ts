// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

export enum CommandPaletteEntities {
    Channel = 'channel',
    Playbooks = 'playbooks',
    Boards = 'boards',
    Messages = 'messages',
    Files = 'files',
    Runs = 'runs',
    GoTo = 'goTo',
}

export enum ChannelType {
    DM_CHANNEL= 'D',
    GM_CHANNEL= 'G',
    OPEN_CHANNEL= 'O',
    PRIVATE_CHANNEL= 'P',
}

export enum BoardsType {
    OPEN_CHANNEL= 'O',
    PRIVATE_CHANNEL= 'P',
}

export enum PlaybooksType {
    OPEN_CHANNEL= 'O',
    PRIVATE_CHANNEL= 'P',
}

