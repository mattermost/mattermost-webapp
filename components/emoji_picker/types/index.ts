// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {EmojiCategory, Emoji} from 'mattermost-redux/types/emojis';

export type Category = {
    className: string;
    emojiIds?: string[];
    id: string;
    message: string;
    name: EmojiCategory;
    offset: number;
};

export type Categories = Record<EmojiCategory, Category>;

export type CategoryOrEmojiRow = {
    type: 'categoryHeaderRow' | 'emojisRow';
    row: EmojiCategory | Emoji[];
}
