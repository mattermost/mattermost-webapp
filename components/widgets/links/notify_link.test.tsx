// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import {ReactWrapper} from 'enzyme';
import {Client4} from 'mattermost-redux/client';

import {act} from 'react-dom/test-utils';

import {StatusOK} from 'mattermost-redux/types/client4';

import {mountWithIntl} from 'tests/helpers/intl-test-helper';

import NotifyLink, {DafaultBtnText} from './notify_link';

const actImmediate = (wrapper: ReactWrapper) =>
    act(
        () =>
            new Promise<void>((resolve) => {
                setImmediate(() => {
                    wrapper.update();
                    resolve();
                });
            }),
    );

describe('components/widgets/links/NotifyLink', () => {
    test('should match snapshot', () => {
        const wrapper = mountWithIntl(
            <NotifyLink/>,
        );
        expect(wrapper).toMatchSnapshot();
    });

    test('should attempt to send and fail', async () => {
        const extraFuncMock = jest.fn();
        const wrapper = mountWithIntl(
            <NotifyLink extraFunc={extraFuncMock}/>,
        );

        const btn = wrapper.find('button');
        expect(btn.text()).toEqual(DafaultBtnText.NotifyAdmin);
        btn.simulate('click');
        expect(btn.text()).toEqual(DafaultBtnText.Sending);

        // wait for the promise called in the try anc catch block to resolve before updating the button text
        await actImmediate(wrapper);

        expect(extraFuncMock).toHaveBeenCalledTimes(1);

        // ultimately fails because the request errors
        expect(btn.text()).toEqual(DafaultBtnText.Failed);
    });

    test('should attempt to send and succeed', async () => {
        const promise = new Promise<StatusOK>(function(resolve) {
            resolve({
                status: 'OK',
            });
        });
        const upgradeRequestMock = jest.spyOn(Client4, 'sendAdminUpgradeRequestEmail');
        upgradeRequestMock.mockImplementation(() => promise);

        const extraFuncMock = jest.fn();

        const wrapper = mountWithIntl(
            <NotifyLink extraFunc={extraFuncMock}/>,
        );

        const btn = wrapper.find('button');
        expect(btn.text()).toEqual(DafaultBtnText.NotifyAdmin);
        btn.simulate('click');
        expect(btn.text()).toEqual(DafaultBtnText.Sending);

        // wait for the promise called in the try anc catch block to resolve before updating the button text
        await actImmediate(wrapper);

        expect(btn.text()).toEqual(DafaultBtnText.Sent);
        expect(extraFuncMock).toHaveBeenCalledTimes(1);
        upgradeRequestMock.mockRestore();
    });
});
