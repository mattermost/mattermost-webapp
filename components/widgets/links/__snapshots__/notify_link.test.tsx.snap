// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/widgets/links/NotifyLink should match snapshot 1`] = `
<NotifyLink>
  <button
    className="btn-link"
    disabled={false}
    onClick={[Function]}
  >
    Notify the admin.
  </button>
</NotifyLink>
`;
